Add controller specs.

fuadmoin opened the pull request:
In this branch I: 

- [ ] wrote a spec for the user controller. 

       -      I tested If the response status was correct for both index and show HTML.
       -      I tested if a correct template was rendered for both index and show HTML.
       -      I tested if the response body includes the correct placeholder text.    

- [ ] wrote a spec for the post controller.

      -      I tested If the response status was correct for both index and show HTML.
      -      I tested if a correct template was rendered for both index and show HTML.
      -      I tested if the response body includes the correct placeholder text.   

- [ ] Updated readme

fuadmoin committed:
Add controller specs.

fuadmoin committed:
Update readme.md

fuadmoin committed:
Update readme.

HAMMAS-SALEEM commented on spec/controllers/posts_controller_spec.rb:
```diff
@@ -0,0 +1,41 @@
+RSpec.describe PostsController, type: :controller do
+  let(:user) { User.create!(name: 'Test User', posts_counter: 0) }
+
+  describe 'GET #index' do
+    it 'assigns the requested user as @user' do
+      get :index, params: { user_id: user.to_param }
+      expect(assigns(:user)).to eq(user)
+    end
+
+    it 'assigns the user posts as @posts' do
+      post = user.posts.create!(title: 'Test Post', likes_counter: 0, comments_counter: 0)
+      get :index, params: { user_id: user.to_param }
+      expect(assigns(:posts)).to eq([post])
+    end
+  end
+
+  describe 'GET #show' do
+    context 'when the post exists' do
+      it 'assigns the requested user as @user' do
+        post = user.posts.create!(title: 'Test Post', likes_counter: 0, comments_counter: 0)
+        get :show, params: { user_id: user.to_param, id: post.to_param }
+        expect(assigns(:user)).to eq(user)
+      end
+
+      it 'assigns the requested post as @post' do
+        post = user.posts.create!(title: 'Test Post', likes_counter: 0, comments_counter: 0)
+        get :show, params: { user_id: user.to_param, id: post.to_param }
+        expect(assigns(:post)).to eq(post)
+      end
+    end
+
+    context 'when the post does not exist' do
+      it 'redirects to the user posts list' do
+        get :show, params: { user_id: user.to_param, id: -1 }
+        expect(response).to redirect_to(user_posts_path(user))
+      end
+    end
+  end
```
| Project Requirements 📝 |
|-|
| ![image](https://github.com/fuadmoin/Blog-app/assets/93039359/601cfa21-c099-44f7-88c1-39b092274cfb) |


- [ ] It would be nice if you follow the project requirements because in the project requirements, it is clearly mentioned that you have to check the things shown in the image above ⬆️ 👍

Hint: You can use the following matchers to complete the testing according to the project requirements 👍

```ruby

--- These below are just examples so you need to write your tests in your own way

RSpec.describe 'Posts', type: :request do
  describe 'GET /index' do
    before(:example) { get '/users/1/posts' }
    it 'renders http status' do
      expect(response).to have_http_status :ok
    end

    it "renders 'index' template" do
      expect(response).to render_template :index
    end

    it 'renders correct placeholder' do
      expect(response.body).to include '<h1>Posts#index</h1>'
    end
  end

  describe 'GET /show' do
    before(:example) { get '/users/1/posts/1' }
    it 'renders http status' do
      expect(response).to have_http_status :ok
    end

    it "renders 'show' template" do
      expect(response).to render_template :show
    end

    it 'renders correct placeholder' do
      expect(response.body).to include '<h1>Single post</h1>'
    end
  end
end
```

HAMMAS-SALEEM commented on spec/controllers/users_controller_spec.rb:
```diff
@@ -0,0 +1,28 @@
+require 'rails_helper'
+
+RSpec.describe UsersController, type: :controller do
+  describe 'GET #index' do
+    it 'assigns all users as @users' do
+      user = User.create!(name: 'John Doe', posts_counter: 0)
+      get :index
+      expect(assigns(:users)).to eq([user])
+    end
+  end
+
+  describe 'GET #show' do
+    context 'when the user exists' do
+      it 'assigns the requested user as @user' do
+        user = User.create!(name: 'John Doe', posts_counter: 0)
+        get :show, params: { id: user.to_param }
+        expect(assigns(:user)).to eq(user)
+      end
+    end
+
+    context 'when the user does not exist' do
+      it 'redirects to the users list' do
+        get :show, params: { id: -1 }
+        expect(response).to redirect_to(users_path)
+      end
+    end
+  end
+end
```
| Project Requirements 📝 |
|-|
| ![image](https://github.com/fuadmoin/Blog-app/assets/93039359/601cfa21-c099-44f7-88c1-39b092274cfb) |


- [ ] It would be nice if you follow the project requirements because, in the project requirements, it is clearly mentioned that you have to check the things shown in the image above ⬆️ 👍

Hint: You can use the following matchers to complete the testing according to the project requirements 👍

```ruby

--- These below are just examples so you need to write your tests in your own way

RSpec.describe 'Users', type: :request do
  describe 'GET /index' do
    before(:example) { get users_path }
    it 'renders http status' do
      expect(response).to have_http_status(200)
    end

    it "renders 'index' template" do
      expect(response).to render_template(:index)
    end

    it 'should return correct placeholder' do
      expect(response.body).to include '<h1>Users#index</h1>'
    end
  end

  describe 'GET /show' do
    before(:example) { get '/users/1' }
    it 'renders http status' do
      expect(response).to have_http_status 200
    end

    it "renders 'show' template" do
      expect(response).to render_template :show
    end

    it 'should return correct placeholder' do
      expect(response.body).to include '<h1>Single User</h1>'
    end
  end
end
```

HAMMAS-SALEEM commented on README.md:
```diff
@@ -127,8 +127,8 @@ rspec spec
 
```
**[OPTIONAL]**
- [ ] It would be nice if you make your pull request summary more detailed to make your pull request more professional and descriptive 👍

HAMMAS-SALEEM reviewed (changes requested):
# Required Changes :recycle:
![Ukhv](https://user-images.githubusercontent.com/93039359/247195854-e274f284-2e3a-423b-8939-7e4efb638ee0.gif)

## Hey @fuadmoin :wave:

## Highlights :1st_place_medal:

✔️ Descriptive PR Title
✔️ Descriptive PR Summary
✔️ Following Gitflow
✔️ Following some requirements
✔️ Great use of tests 👍
✔️ Tests are passing 😊

Good job on implementing some of the requirements, there are some issues that you still need to work on to go to the next project but you are almost there! :blush:

# Optional

_All the changes with optional tags are not crucial enough to prevent you from getting the approval but we highly recommend to students that they should take these changes into consideration._

Happy coding! :clap: :clap: :clap:

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

---------------------------------------

_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

fuadmoin committed:
Make required changes.

I re-wrote the tests based on the given suggestion and the required changes, i.e checked - if the response includes a placeholder text. - render the templates. - returns a successful response.

mikethreels reviewed (approved):
Hi Fuad, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
- Good job following the instructions of the previous CR.
## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
